### PR TITLE
Feature update namespace

### PR DESCRIPTION
**Pull request by ocallesp**
This PR adds a feature where after moving compile items between folders (by drag drop on Solution Explorer), the user is prompted to update the namespaces.

Requires VS PR: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/CPS/pullrequest/378467
Requires CPS PR: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/CPS/pullrequest/378467

Two new options are added to control this behaviour:

image

image

image

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7925)

### CODE REVIEW

File: build/import/Packages.targets [hunk 1]
```diff
@@ -60,23 +61,23 @@
...
     <!-- https://github.com/dotnet/roslyn/issues/53877 -->
-    <PackageReference Include="StreamJsonRpc"                                                         Version="2.8.28" />
+    <PackageReference Include="StreamJsonRpc"                                                         Version="2.10.42" />
 
     <!-- CPS -->
-    <PackageReference Update="Microsoft.VisualStudio.ProjectSystem"                                   Version="17.1.221-pre" />
-    <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.Analyzers"                         Version="17.1.221-pre" />
-    <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.Query"                             Version="17.1.221-pre" />
-    <PackageReference Update="Microsoft.VisualStudio.ProjectSystem.SDK.Tools"                         Version="17.1.221-pre" />
+    <PackageReference Update="Microsoft.VisualStudio.ProjectSystem"                                   Version="17.2.192-pre-gd98aebaca6" />
```
Comment (ocallesp) [inline build/import/Packages.targets:77]: This CPS package version are for testing. At the end I will put the latest CPS versions.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveActionHandler.cs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        private readonly Workspace _workspace;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IRoslynServices _roslynServices;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IProjectAsynchronousTasksService _projectAsynchronousTasksService;
+        private readonly IWaitIndicator _waitService;
+        private readonly IUnconfiguredProjectVsServices _unconfiguredProjectVsServices;
+
+        [ImportingConstructor]
+        public FileMoveActionHandler(IFileMoveWithinProjectNotifier fileMoveWithinProjectNotifier,
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveActionHandler.cs:39]: Did you deploy CPS to your exp hive too?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveHandler.cs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                string filename = Path.GetFileName(filenameWithPath);
+
+                var oldDocument = project.Documents.FirstOrDefault(d => StringComparers.Paths.Equals(d.FilePath, filenameWithPath));
+                if (oldDocument is null)
+                {
+                    continue;
+                }
+
+                // This is a file move to another directory, it should only detect this a Update Namespace action.
+                Renamer.RenameDocumentActionSet documentAction = await Renamer.RenameDocumentAsync(oldDocument, filename, new List<string>(new string[] { destinationFolder }));
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveHandler.cs:172]: This creates more allocations than needed:

```c#
new List<string>(new string[] { destinationFolder })
```

Perhaps just

```c#
new List<string> { destinationFolder }
```

or (better, if it works):

```c#
new string[] { destinationFolder }
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/VSResources.resx [hunk 4]
```diff
@@ -293,4 +293,7 @@ In order to debug this project, add an executable project to this solution which
     <value>WARNING: Potential incremental build failure in '{0}'. See: https://aka.ms/incremental-build-failure</value>
     <comment>{0} is the file name of the project.</comment>
   </data>
+  <data name="UpdateNamespacePromptMessage" xml:space="preserve">
+    <value>Adjust namespaces for moved files ?</value>
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/VSResources.resx:297]: Do we have guidance from any PMs on what this text should be? cc @mckennabarlow

**Review by drewnoakes — COMMENTED** [review]
Just some quick feedback during a brief scan. I'll take a more detailed look next week.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**